File: app/apps/u7406/lib/actions/graph/query/edges.ex
defmodule Actions.Graph.Query.Edges do
  import U7406

  alias Queries.Graph
  alias Schemas.Graph.Division
  alias U7406.Repo

  def run(_parent, args, _context) do
    %{q: q, division_id: id, offset: offset, limit: limit} = args

    parameters = Jason.decode!(q)

    query =
      Repo.get(Division, id)
      |> assoc(:edges)
      |> Graph.join_assocs([:edge_type, :start_node, :end_node])
      |> Graph.search(parameters)

    edges = query |> Graph.paginate(offset, limit) |> Repo.all()
    total = query |> Repo.aggregate(:count)

    {:ok,
     %{
       entries: edges,
       offset: offset,
       limit: limit,
       total: total
     }}
  end
end

File: mgt/src/graphql/query/Edges.ts
const QUERY = /* GraphQL */ `
  query Edges($divisionId: ID!, $q: String, $offset: Int, $limit: Int) {
    edges(divisionId: $divisionId, q: $q, offset: $offset, limit: $limit) {
      entries {
        id
        name
        edgeTypeId
        props {
          key
          val
        }
        edgeType {
          id
          name
          uid
        }
        startNode {
          id
          name
          uid
        }
        endNode {
          id
          name
          uid
        }
      }
      limit
      offset
      total
    }
  }
`
export default QUERY
